openstack_hypervisor/fileserver/server.py

```diff
@@ -0,0 +1,400 @@
+compression), and file downloads. TLS support is configured via
+``oslo_service.sslutils`` and ``oslo_config``.
+"""
+
+import json
+import os
+import ssl
+import threading
+import uuid
+from pathlib import Path
+from socketserver import ThreadingMixIn
+from typing import Any, Dict
+from wsgiref.simple_server import WSGIServer, make_server
+
+from oslo_config import cfg
+from oslo_log import log as logging
+from oslo_service import service, sslutils
+from webob import Request, Response
+
+from .utils import (
+    CHUNK_READ_SIZE,
+    BadRequestError,
+    assemble_chunks_to_file,
+    atomic_replace,
+    cleanup_upload,
+    compute_sha256,
+    get_snap_common,
+    read_meta_json,
+    uploads_root,
+    write_meta_json,
+)
+
+LOG = logging.getLogger(__name__)
+
+
+fileserver_opts = [
+    cfg.StrOpt(
+        "host",
+        default=os.environ.get("FILESERVER_HOST", "0.0.0.0"),
+        help="Listen address for the file server",
+    ),
+    cfg.IntOpt(
+        "port",
+        default=int(os.environ.get("FILESERVER_PORT", "8099")),
+        min=1,
+        help="TCP listen port for the file server",
+    ),
+]
+
+CONF = cfg.CONF
+CONF.register_opts(fileserver_opts, group="fileserver")
+sslutils.register_opts(CONF)
+
+
+def _json(request: Request) -> Dict[str, Any]:
+    """Parse JSON body into a dictionary.
+
+    Returns an empty dict when the body is empty. Raises BadRequestError on
+    invalid JSON.
+    """
+    if not request.body_file:
+        return {}
+    try:
+        body = request.body
+    except Exception:
+        body = request.body_file.read()
+    if not body:
+        return {}
+    try:
+        return json.loads(body.decode("utf-8"))
+    except Exception as exc:
+        raise BadRequestError(f"invalid JSON body: {exc}")
+
+
+def _ok(payload: Dict[str, Any]) -> Response:
+    """Return a JSON 200 OK response with the provided payload."""
+    return Response(json_body=payload)
+
+
+def _error(status: int, detail: str) -> Response:
+    """Return a JSON error response with the given HTTP status and detail."""
+    return Response(json_body={"detail": detail}, status=status)
+
+
+def create_file_ep(request: Request) -> Response:
+    """Create an empty file at the requested destination path."""
+    try:
+        payload = _json(request)
+        dst_path = payload.get("dst_path")
+        if not dst_path:
+            raise BadRequestError("dst_path is required")
+        dst = Path(dst_path)
+        dst.parent.mkdir(parents=True, exist_ok=True)
+        try:
+            with open(dst, "xb"):
+                pass
+        except FileExistsError:
+            pass
+        return _ok({"created": True, "path": str(dst)})
+    except BadRequestError as exc:
+        return _error(400, str(exc))
+    except Exception as exc:
+        LOG.exception("create_file failed: %s", exc)
+        return _error(500, str(exc))
+
+
+def remove_file_ep(request: Request) -> Response:
```

Comment:
create file and remove file should operate within a sandbox path, otherwise we are leaving a door open (potentially) to create/remove files arbitrarily in the filesystem (within the limits of the snap and/or the user this service is running)